kpi-tracker/src/main.rs

```diff
@@ -702,30 +765,131 @@ async fn use_node(db: &mut DB, from_height: AbsoluteBlockHeight) -> anyhow::Resu
         .context("Could not connect to node.")?;
 
     let mut blocks_stream = node
-        .get_finalized_blocks_from(from_height)
+        .get_finalized_blocks_from(latest_height)
         .await
         .context("Error querying blocks")?;
 
-    let start_height = if from_height.height == 0 {
+    if latest_height.height == 0 {
         if let Some(genesis_block) = blocks_stream.next().await {
-            process_genesis_block(&mut node, genesis_block.block_hash, db).await?;
+            let genesis_block_data =
+                process_genesis_block(&mut node, genesis_block.block_hash).await?;
+            block_sender
+                .send(BlockData::Genesis(genesis_block_data))
+                .await?;
             log::info!("Processed genesis block: {}", genesis_block.block_hash);
         }
-        from_height.height + 1
-    } else {
-        from_height.height
     };
 
-    for height in start_height..blocks_to_process {
+    for height in latest_height.height.. {
         if let Some(block) = blocks_stream.next().await {
-            process_block(&mut node, block.block_hash, db).await?;
+            let block_data = process_block(&mut node, block.block_hash).await?;
+            block_sender.send(BlockData::Normal(block_data)).await?;
             log::info!("Processed block ({}): {}", height, block.block_hash);
         }
     }
 
     Ok(())
 }
 
+/// Inserts the `block_data` collected for a single block into the database
+/// defined by `db`. Everything is commited as a single transactions allowing
+/// for easy restoration from the last recorded block (by height) inserted into
+/// the database.
+async fn db_insert_block(db: &mut DBConn, block_data: BlockData) -> anyhow::Result<()> {
+    let db_tx = db
+        .client
+        .transaction()
+        .await
+        .context("Failed to build DB transaction")?;
+
+    let tx_ref = &db_tx;
+    let prepared_ref = &db.prepared;
+
+    let insert_common = |block_hash: BlockHash, block_details: BlockDetails, accounts: Accounts| async move {
+        let block_id = prepared_ref
+            .insert_block(tx_ref, block_hash, block_details)
+            .await?;
+
+        for (address, details) in accounts.into_iter() {
```

Comment:
Please do not use hashmaps. This iteration here has unspecified order which means we have nondeterminism where we don't need to have nondeterminism.